orderbook/src/orderbook.rs

```diff
@@ -89,74 +63,65 @@ impl Orderbook {
     }
 }
 
-pub fn has_future_valid_to(now_in_epoch_seconds: u32, order: &OrderCreation) -> bool {
+fn has_future_valid_to(now_in_epoch_seconds: u32, order: &OrderCreation) -> bool {
     order.valid_to > now_in_epoch_seconds
 }
 
+async fn set_order_balance(orders: &mut [Order], balance_fetcher: &dyn BalanceFetching) {
+    // Since order can come from storage after a cold start there is the possibility that they are not yet registered
+    // for balance updates. In this case we do it here.
+    let untracked = orders
+        .iter()
+        .filter_map(|order| {
+            match balance_fetcher
+                .get_balance(order.order_meta_data.owner, order.order_creation.sell_token)
+            {
+                Some(_) => None,
+                None => Some((order.order_meta_data.owner, order.order_creation.sell_token)),
+            }
+        })
+        .collect();
+    balance_fetcher.register_many(untracked).await;
+
+    // Enrich orders with balance information
+    for order in orders.iter_mut() {
+        order.order_meta_data.available_balance = balance_fetcher
+            .get_balance(order.order_meta_data.owner, order.order_creation.sell_token);
+    }
+}
+
+fn remove_orders_without_sufficient_balance(orders: &mut Vec<Order>) {
+    orders.retain(|order| {
+        let balance = order.order_meta_data.available_balance.unwrap_or_default();
+        !balance.is_zero()
+            && (order.order_creation.partially_fillable
+                || balance >= order.order_creation.sell_amount)
+    });
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::{account_balances::MockBalanceFetching, storage::MockStorage};
+    use crate::account_balances::MockBalanceFetching;
     use ethcontract::H160;
-    use mockall::{
-        predicate::{always, eq},
-        Sequence,
-    };
-    use model::{
-        order::{OrderBuilder, OrderCreation},
-        DomainSeparator,
-    };
-
-    #[tokio::test]
-    async fn watches_owners_sell_token_balance_for_added_orders() {
```

Comment:
This test I removed because all it tests is that add_order calls balance_fetcher.register which we do unconditionally so I don't think there is a point in having a test for it.

Reply:
Even if the codepath is unconditional, I feel like it's a nice documentation and assurance of behaviour of the public contract of this component. If one day we refactored the orderbook and remove the line to register balances for on the add path I think it's nice when an automated test catches that.

That being said, if this is the only thing requiring the mock, I'm ok with letting it go.